carma-messenger-core/j2735_convertor/src/psm_convertor.cpp

```diff
@@ -108,52 +113,104 @@ void PSMConvertor::convert(const j2735_v2x_msgs::msg::AttachmentRadius& in_msg,
     out_msg.attachment_radius = in_msg.attachment_radius;
 };
 
-
 void PSMConvertor::convert(const j2735_v2x_msgs::msg::Heading& in_msg, carma_v2x_msgs::msg::Heading& out_msg) {
-    out_msg.heading = in_msg.heading * heading_conversion_const_;
+        out_msg.heading = in_msg.heading * heading_conversion_const_;
 };
 
 void PSMConvertor::convert(const carma_v2x_msgs::msg::Heading& in_msg, j2735_v2x_msgs::msg::Heading& out_msg) {
-    out_msg.heading = in_msg.heading / heading_conversion_const_;
+        out_msg.heading = in_msg.heading / heading_conversion_const_;
 };
 
 void PSMConvertor::convert(const j2735_v2x_msgs::msg::Velocity& in_msg, carma_v2x_msgs::msg::Velocity& out_msg) {
-    out_msg.velocity = in_msg.velocity * velocity_conversion_const_;
+        out_msg.velocity = in_msg.velocity * velocity_conversion_const_;
 };
 
 void PSMConvertor::convert(const carma_v2x_msgs::msg::Velocity& in_msg, j2735_v2x_msgs::msg::Velocity& out_msg) {
-    out_msg.velocity = in_msg.velocity / velocity_conversion_const_;
+        out_msg.velocity = in_msg.velocity * velocity_conversion_const_;
 };
 
 
 void PSMConvertor::convert(const j2735_v2x_msgs::msg::PSM& in_msg, carma_v2x_msgs::msg::PSM& out_msg)
 {
+    if(in_msg.HAS_ACCEL_SET) {
       convert(in_msg.accel_set, out_msg.accel_set);
+    }
+
+    if(in_msg.HAS_PATH_PREDICTION) {
       convert(in_msg.path_prediction, out_msg.path_prediction);
-      convert(in_msg.accuracy, out_msg.accuracy);
-      convert(in_msg.position, out_msg.position);
-      convert(in_msg.path_history, out_msg.path_history);
-      convert(in_msg.attachment_radius, out_msg.attachment_radius);
-      convert(in_msg.heading, out_msg.heading);
-      convert(in_msg.speed, out_msg.speed);
+    }
 
-      out_msg.propulsion = in_msg.propulsion;
-      out_msg.use_state = in_msg.use_state;
-      out_msg.cross_request = in_msg.cross_request;
-      out_msg.cross_state = in_msg.cross_state;
+    convert(in_msg.accuracy, out_msg.accuracy);
+    convert(in_msg.position, out_msg.position);
+
+        convert(in_msg.heading, out_msg.heading);
+        convert(in_msg.speed, out_msg.speed);
+
+    if(in_msg.HAS_PATH_HISTORY) {
```

Comment:
This should be (in_msg.presence_vector & j2735_v2x_msgs::msg::PSM::HAS_PATH_HISTORY)
same for all instances where we are checking if the field exists